README.md

```diff
@@ -1,16 +1,1261 @@
+>
+> — [Mark Miller](https://github.com/erights), [Robust Composition](http://www.erights.org/talks/thesis/markm-thesis.pdf)
+
+There MAY not be enough information to described an Invocation at creation time. However, all of the information required to construct the next request in a sequence MAY be available in the same Batch, or in a previous (but not yet complete) Invocation. Waiting for each request to complete before proceeding to the next task has a performance impact due to the amount of latency. [Promise pipelining](http://erights.org/elib/distrib/pipeline.html) is a solution to this problem: by referencing a prior invocation, a pipelined invocation can direct the executor to use the output of earlier invocations into the input of a later one. This liberates the invoker from waiting for each step.
+
+A Promise is a placeholder value MAY be used as a variable placeholder for a concrete value in a [Closure](#3-closure), waiting on a previous step to complete.
+
+One way of seeing the names in a [`Batch`](#5-batch) is as variables for the return of each Closure. These can now be referenced by other Closures.
+
+For example, consider the following batch:
+
+``` json
+{
+  "run": {
+    "create-draft": {
+      "with": "https://example.com/blog/posts",
+      "do": "crud/create",
+      "inputs": {
+        "payload": {
+          "title": "How UCAN Tasks Changed My Life",
+          "body": "This is the story of how one spec changed everything..."
+        }
+      }
+    },
+    "get-editors": {
+      "with": "https://example.com/users/editors",
+      "do": "crud/read"
+    },
+    "notify": {
+      "with": "mailto:akiko@example.com",
+      "do": "msg/send",
+      "inputs": {
+        "to": {"ucan/ok": ["/", "right"]},
+        "subject": "Coffee",
+        "body": {"ucan/ok": ["/", "left"]}
+      }
+    }
+  }
+}
+```
+
+By analogy, this can be interpreted roughly as follows:
+
+``` js
+const createDraft = crud.create("https://example.com/blog/posts", {
+  payload: {
+    title: "How UCAN Tasks Changed My Life",
+    body: "This is the story of how one spec changed everything..."
+  }
+})
+
+const getEditors = crud.read("https://example.com/users/editors")
+
+const notify = msg.send("mailto:akiko@example.com", {
+  "to": await createDraft,
+  "subject": "Coffee",
+  "body": await getEditors
+})
+```
+
+While a Promise MAY be substituted for any field in a Closure, substituting the `do` field is NOT RECOMMENDED. The `do` field is critical in understanding what kind of action will be performed, and most schedulers will use the 
+
+After resolution, the Closure MUST be validated against the UCANs known to the Executor. A Promise resolved to a Closure that is not backed by a valid UCAN MUST NOT be executed, and SHOULD return an unauthorized error to the user.
```

Comment:
I think returned closure or task should never imply invoking it, it pass of reference not an intention to invoke it.

Reply:
The purpose of putting something inside the Invocation is to say "please run this thing". If they want to make that decision, it should either be captured as e.g. a Wasm step that can handle the domain logic, or returned to the caller who can decide what to do in the next invocation.

Reply:
I changed working to Task, but there is always an implied execution request by coming wrapped in an invocation. You could construct a runtime that makes a decision, but then it either shouldn't be part of an existing invocation, or have some logic written in e.g. Wasm that decides on which side of a branch to push the dataflow into.